lib/ting/hanyu_pinyin_parser.rb

```diff
@@ -15,7 +15,15 @@ def all_syllables
     end
 
     def sylls_with_erhua
-      @with_erhua ||= all_syllables.map{|p| p + 'r'}
+      @with_erhua ||= begin
+        with_erhua = all_syllables.dup
+        # We assume that syllables ending in 'r' cannot have additional erhua.
+        with_erhua.reject!{|p| p.end_with? "r"}
+        # Don't shadow existing syllables, to avoid parsing "er" as "e" + "er".
+        with_erhua.reject!{|p| all_syllables.include? p + 'r'}
+        # Erhua syllables must be followed by whitespace.
+        with_erhua.map{|p| p + 'r '}
```

Comment:
Instead of using destructive `reject!` calls, and then having to defensively `dup` the input, I would just chain these calls, that way you also can get rid of the begin/end and the extra local variable.

``` ruby
@with_erhua ||= all_syllables
                  .reject {|p| p.end_with? "r"}                 # We assume that syllables ending in 'r' cannot have additional erhua.
                  .reject {|p| all_syllables.include? p + 'r'}  # Don't shadow existing syllables, to avoid parsing "er" as "e" + "er".
                  .map {|p| p + 'r '}                           # Erhua syllables must be followed by whitespace.
```

Reply:
I'm not a fan of destructive operations either, but I also didn't want to introduce super long lines with chaining + comments. What do you think of trailing dots and comments in-between? (Just pushed it)